block-producer/src/block_builder/kernel.rs

```diff
@@ -0,0 +1,239 @@
+
+        Self {
+            program: account_program,
+        }
+    }
+
+    // Note: this will eventually all be done in the VM
+    pub fn compute_block_header(
+        &self,
+        witness: BlockHeaderWitness,
+    ) -> Result<BlockHeader, BuildBlockError> {
+        let prev_hash = witness.prev_header.prev_hash();
+        let block_num = witness.prev_header.block_num();
+        let version = witness.prev_header.version();
+
+        let chain_root = Digest::default();
+        let account_root = self.compute_new_account_root(witness)?;
+        let nullifier_root = Digest::default();
+        let note_root = Digest::default();
+        let batch_root = Digest::default();
+        let proof_hash = Digest::default();
+        let timestamp: Felt = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("today is expected to be before 1970")
+            .as_millis()
+            .into();
+
+        Ok(BlockHeader::new(
+            prev_hash,
+            block_num,
+            chain_root,
+            account_root,
+            nullifier_root,
+            note_root,
+            batch_root,
+            proof_hash,
+            version,
+            timestamp,
+        ))
+    }
+
+    /// `current_account_states`: iterator of (account id, node hash, Merkle path)
+    /// `account_updates`: iterator of (account id, new account hash)
+    fn compute_new_account_root(
+        &self,
+        witness: BlockHeaderWitness,
+    ) -> Result<Digest, BlockKernelError> {
+        let (advice_inputs, stack_inputs) = witness.into_parts()?;
+        let host = {
+            let advice_provider = MemAdviceProvider::from(advice_inputs);
+
+            DefaultHost::new(advice_provider)
+        };
+
+        let execution_output =
+            execute(&self.program, stack_inputs, host, ExecutionOptions::default())
+                .map_err(BlockKernelError::ProgramExecutionFailed)?;
+
+        let new_account_root = {
+            let stack_output = execution_output.stack_outputs().stack_truncated(4);
+
+            let digest_elements: Vec<Felt> = stack_output
+            .iter()
+            .map(|&num| Felt::try_from(num).map_err(|_|BlockKernelError::InvalidRootReturned))
+            // We reverse, since a word `[a, b, c, d]` will be stored on the stack as `[d, c, b, a]`
+            .rev()
+            .collect::<Result<_, BlockKernelError>>()?;
+
+            let digest_elements: [Felt; 4] =
+                digest_elements.try_into().map_err(|_| BlockKernelError::InvalidRootReturned)?;
+
+            digest_elements.into()
+        };
+
+        Ok(new_account_root)
+    }
+}
+
+/// Provides inputs to the `BlockKernel` so that it can generate the new header
+pub(super) struct BlockHeaderWitness {
+    account_states: Vec<AccountInputRecord>,
+    account_updates: Vec<(AccountId, Digest)>,
+
+    /// Note: this field will no longer be necessary when we have a full block kernel.
+    prev_header: BlockHeader,
```

Comment:
I think we might still need `prev_header` - so, not sure the comment is accurate.